folium/features.py

```diff
@@ -409,11 +409,13 @@ def __init__(self, data, style_function=None, name=None,
             else:  # This is a filename
                 with open(data) as f:
                     self.data = json.loads(f.read())
-        elif hasattr(data, '__geo_interface__'):
+        elif hasattr(data, 'geometry'):
```

Comment:
`__geo_interface__` takes any object that implements the `__geo_interface__` but `to_json()` is `geopandas` only.

Reply:
GeoDataFrames that have any rows with missing geometries don't implement `__geo_interface__`.

So should this `elif` block be kept the same and have an additional block, `elif hasattr(data, 'geometry') and any(data['geometry'].isnull()):`?

Reply:
Interesting, so something is wrong with that logic b/c then you should be directed to the `else` and things should fail as expected, no? If that is the case all you have to do is the improve the error message in the `else`.

Reply:
Right, but the idea is to have an error message specifically for this case, and leave the existing, vague error message for other cases that fail.

Reply:
Ah, okay. So there are no reasons besides missing geometries that you could get to the `else` block in the current code?

Reply:
> Ah, okay. So there are no reasons besides missing geometries that you could get to the `else` block in the current code?

That is not what I said, but in a way any "bad" object does not have a `geometry`, so the message are equivalent. One better way to phrase that would be:

`Could not find valid geometries in {!r}`, so the object `__repr__` is printed in the error.

Reply:
I just think that people using GeoDataFrames that have, say, a single row with a missing geometry, might be confused by that, and would appreciate a more descriptive message.

Reply:
I believe that the error message above is descriptive enough but feel free to improve on it. However, the current if-clauses would kill all objects that we accept via the geo interface and that is a no-no.